lib/php/rendition-factory/src/Transformer/Video/FFMpegTransformerModule.php

```diff
@@ -20,39 +23,79 @@ public static function getName(): string
+            if (!in_array($videoCodec, $ouputFormat->getAvailableVideoCodecs())) {
+                throw new \InvalidArgumentException(sprintf('Invalid video codec %s for format %s', $videoCodec, $format));
+            }
+            $ouputFormat->setVideoCodec($videoCodec);
+        }
+        if ($audioCodec = $options['audio_codec'] ?? null) {
+            if (!in_array($audioCodec, $ouputFormat->getAvailableAudioCodecs())) {
+                throw new \InvalidArgumentException(sprintf('Invalid audio codec %s for format %s', $audioCodec, $format));
+            }
+            $ouputFormat->setAudioCodec($audioCodec);
+        }
+
+        $ffmpeg = FFMpeg\FFMpeg::create(); // (new FFMpeg\FFMpeg)->open('/path/to/video');
+        /** @var Video $video */
+        $video = $ffmpeg->open($inputFile->getPath());
+
+        $filters = $options['filters'] ?? [];
+        // first, turn the video into a clip
+        if (!empty($filters) && 'pre_clip' === $filters[0]['name']) {
+            $filter = array_shift($filters);
+            $context->getLogger()->info(sprintf('Applying filter: %s', $filter['name']));
+            $clip = $this->pre_clip($video, $filter, $context);
+        } else {
+            $clip = $video->clip(TimeCode::fromSeconds(0), TimeCode::fromString('01:00:00:00.00'));
         }
-        /** @var FFMpeg\Format\VideoInterface $ouputFormat */
-        $outputFormat = new $format();
 
-        $video = $ffmpeg->open( $inputFile->getPath());
-        $videoFiltered = $video->filters();
-        foreach($options['filters']??[] as $filter) {
-            if(!method_exists($this, $filter['name'])) {
+        foreach ($filters as $filter) {
+            if ('pre_clip' === $filter['name']) {
+                throw new \InvalidArgumentException('"pre_clip" filter must be the first filter');
+            }
+            if (!method_exists($this, $filter['name'])) {
                 throw new \InvalidArgumentException(sprintf('Invalid filter: %s', $filter['name']));
             }
             $context->log(sprintf('Applying filter: %s', $filter['name']));
 
-            /** @uses self::resize(), self::rotate(), self::pad(), self::crop(), self::clip(), self::synchronize()
-             *  @uses self::watermark(), self::framerate()
+            /* @uses self::resize(), self::rotate(), self::pad(), self::crop(), self::clip(), self::synchronize()
+             *  @uses self::watermark(), self::framerate(), self::remove_audio()
              */
-            $videoFiltered = call_user_func([$this, $filter['name']], $videoFiltered, $filter, $context);
+            call_user_func([$this, $filter['name']], $clip, $filter, $context);
         }
 
         $outputPath = $context->createTmpFilePath($extension);
 
-        $video->save($outputFormat, $outputPath);
+        // $video->save($ouputFormat, $outputPath);
+        $clip->save($ouputFormat, $outputPath);
 
+        unset($clip);
         unset($video);
 
         return new OutputFile(
```

Comment:
OutputFile should return a known family type

Reply:
yes but not so easy because the output extension will change the "family" of the output file (.mp4 -> video ; .jpg -> image ; .wav -> audio...)
:( we need to implement a strong mapping.

Reply:
Yes but this is mandatory, otherwise we cannot chain modules...